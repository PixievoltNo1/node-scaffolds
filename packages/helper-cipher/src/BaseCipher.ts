import type ChalkLogger from '@guanghechen/chalk-logger'
import { mkdirsIfNotExists } from '@guanghechen/helper-path'
import { consumeStream, consumeStreams, destroyBuffers } from '@guanghechen/helper-stream'
import type { Cipher } from 'crypto'
import fs from 'fs-extra'
import type { ICipher } from './types/cipher'

/**
 * ICipher base class.
 */
export abstract class BaseCipher implements ICipher {
  protected readonly logger?: ChalkLogger

  constructor(logger?: ChalkLogger) {
    this.logger = logger
  }

  // override
  public encrypt(plainData: Readonly<Buffer>): Buffer | never {
    const encipher = this.encipher()
    const cipherDataPieces: Buffer[] = []
    let cipherData: Buffer

    try {
      // Collect and encrypt data
      cipherDataPieces.push(encipher.update(plainData))
      cipherDataPieces.push(encipher.final())
      cipherData = Buffer.concat(cipherDataPieces)
    } finally {
      destroyBuffers(cipherDataPieces)
      encipher.destroy()
    }

    return cipherData
  }

  // override
  public decrypt(cipherData: Readonly<Buffer>): Buffer | never {
    const decipher = this.decipher()
    const plainDataPieces: Buffer[] = []
    let plainData: Buffer

    try {
      // Collect and decrypt data
      plainDataPieces.push(decipher.update(cipherData))
      plainDataPieces.push(decipher.final())
      plainData = Buffer.concat(plainDataPieces)
    } finally {
      destroyBuffers(plainDataPieces)
      decipher.destroy()
    }

    return plainData
  }

  // override
  public async encryptFromFiles(plainFilepaths: string[]): Promise<Buffer | never> {
    const encipher: Cipher = this.encipher()
    const pieces: Buffer[] = []
    try {
      const readers: NodeJS.ReadableStream[] = plainFilepaths.map(fp => fs.createReadStream(fp))
      await consumeStreams(readers, encipher)
      for await (const chunk of encipher) pieces.push(chunk)
      return Buffer.concat(pieces)
    } finally {
      destroyBuffers(pieces)
      encipher.destroy()
    }
  }

  // override
  public async decryptFromFiles(cipherFilepaths: string[]): Promise<Buffer> {
    const decipher: Cipher = this.decipher()
    const pieces: Buffer[] = []
    try {
      const readers: NodeJS.ReadableStream[] = cipherFilepaths.map(fp => fs.createReadStream(fp))
      await consumeStreams(readers, decipher)
      for await (const chunk of decipher) pieces.push(chunk)
      return Buffer.concat(pieces)
    } finally {
      destroyBuffers(pieces)
      decipher.destroy()
    }
  }

  // @override
  public encryptFile(plainFilepath: string, cipherFilepath: string): Promise<void> {
    mkdirsIfNotExists(cipherFilepath, false, this.logger)
    const reader: NodeJS.ReadableStream = fs.createReadStream(plainFilepath)
    const writer: NodeJS.WritableStream = fs.createWriteStream(cipherFilepath)
    const encipher: Cipher = this.encipher()
    return consumeStream(reader, writer, encipher)
  }

  // @override
  public decryptFile(cipherFilepath: string, plainFilepath: string): Promise<void> {
    mkdirsIfNotExists(plainFilepath, false, this.logger)
    const reader: NodeJS.ReadableStream = fs.createReadStream(cipherFilepath)
    const writer: NodeJS.WritableStream = fs.createWriteStream(plainFilepath)
    const decipher: Cipher = this.decipher()
    return consumeStream(reader, writer, decipher)
  }

  // override
  public async encryptFiles(plainFilepaths: string[], cipherFilepath: string): Promise<void> {
    if (plainFilepaths.length <= 0) return
    if (plainFilepaths.length === 1) {
      await this.encryptFile(plainFilepaths[0], cipherFilepath)
      return
    }

    mkdirsIfNotExists(cipherFilepath, false, this.logger)
    const readers: NodeJS.ReadableStream[] = plainFilepaths.map(fp => fs.createReadStream(fp))
    const writer: NodeJS.WritableStream = fs.createWriteStream(cipherFilepath)
    const encipher: Cipher = this.encipher()
    await consumeStreams(readers, writer, encipher)
    encipher.destroy()
  }

  // override
  public async decryptFiles(cipherFilepaths: string[], plainFilepath: string): Promise<void> {
    if (cipherFilepaths.length <= 0) return
    if (cipherFilepaths.length === 1) {
      await this.decryptFile(cipherFilepaths[0], plainFilepath)
      return
    }

    mkdirsIfNotExists(plainFilepath, false, this.logger)
    const readers: NodeJS.ReadableStream[] = cipherFilepaths.map(fp => fs.createReadStream(fp))
    const writer: NodeJS.WritableStream = fs.createWriteStream(plainFilepath)
    const decipher: Cipher = this.decipher()
    await consumeStreams(readers, writer, decipher)
    decipher.destroy()
  }

  protected abstract encipher(): Cipher

  protected abstract decipher(): Cipher

  // @override
  public abstract createSecret(): Buffer

  // @override
  public abstract initFromSecret(secret: Readonly<Buffer>): void | never

  // @override
  public abstract initFromPassword(password: Readonly<Buffer>): void | never

  // @override
  public abstract cleanup(): void
}
