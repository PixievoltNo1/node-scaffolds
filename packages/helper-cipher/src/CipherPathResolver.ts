import { absoluteOfWorkspace, relativeOfWorkspace } from '@guanghechen/helper-path'

export interface ICipherPathResolverProps {
  /**
   * Root directory of source files.
   */
  readonly sourceRootDir: string

  /**
   * Root directory of target files.
   */
  readonly targetRootDir: string
}

export class CipherPathResolver {
  /**
   * Root directory of source files.
   */
  public readonly sourceRootDir: string

  /**
   * Root directory of target files.
   */
  public readonly targetRootDir: string

  constructor(props: ICipherPathResolverProps) {
    this.sourceRootDir = props.sourceRootDir
    this.targetRootDir = props.targetRootDir
  }

  /**
   * Resolve the absolute path of a source file.
   * @param sourceFilepath
   */
  public calcAbsoluteSourceFilepath(sourceFilepath: string): string {
    const filepath = absoluteOfWorkspace(this.sourceRootDir, sourceFilepath)
    return filepath
  }

  /**
   * Resolve the absolute path of a target file.
   * @param targetFilepath
   */
  public calcAbsoluteTargetFilepath(targetFilepath: string): string {
    const filepath = absoluteOfWorkspace(this.targetRootDir, targetFilepath)
    return filepath
  }

  /**
   * Resolve the relative path of the source file.
   * @param absoluteSourceFilepath
   */
  public calcRelativeSourceFilepath(absoluteSourceFilepath: string): string {
    const filepath = relativeOfWorkspace(this.sourceRootDir, absoluteSourceFilepath)
    return filepath.replace(/[/\\]+/g, '/')
  }

  /**
   * Resolve the relative path of a target file.
   * @param absoluteTargetFilepath
   */
  public calcRelativeTargetFilepath(absoluteTargetFilepath: string): string {
    const filepath = relativeOfWorkspace(this.targetRootDir, absoluteTargetFilepath)
    return filepath.replace(/[/\\]+/g, '/')
  }
}
