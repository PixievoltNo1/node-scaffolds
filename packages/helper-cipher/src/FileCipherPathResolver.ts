import { absoluteOfWorkspace, relativeOfWorkspace } from '@guanghechen/helper-path'

export interface IFileCipherPathResolverProps {
  /**
   * Root directory of source files.
   */
  readonly sourceRootDir: string

  /**
   * Root directory of encrypted files.
   */
  readonly encryptedRootDir: string
}

export class FileCipherPathResolver {
  /**
   * Root directory of source files.
   */
  public readonly sourceRootDir: string

  /**
   * Root directory of encrypted files.
   */
  public readonly encryptedRootDir: string

  constructor(props: IFileCipherPathResolverProps) {
    this.sourceRootDir = props.sourceRootDir
    this.encryptedRootDir = props.encryptedRootDir
  }

  /**
   * Resolve the absolute path of a source file.
   * @param sourceFilepath
   */
  public calcAbsoluteSourceFilepath(sourceFilepath: string): string {
    const filepath = absoluteOfWorkspace(this.sourceRootDir, sourceFilepath)
    return filepath
  }

  /**
   * Resolve the absolute path of a encrypted file.
   * @param encryptedFilepath
   */
  public calcAbsoluteEncryptedFilepath(encryptedFilepath: string): string {
    const filepath = absoluteOfWorkspace(this.encryptedRootDir, encryptedFilepath)
    return filepath
  }

  /**
   * Resolve the relative path of the source file.
   * @param absoluteSourceFilepath
   */
  public calcRelativeSourceFilepath(absoluteSourceFilepath: string): string {
    const filepath = relativeOfWorkspace(this.sourceRootDir, absoluteSourceFilepath)
    return filepath.replace(/[/\\]+/g, '/')
  }

  /**
   * Resolve the relative path of a encrypted file.
   * @param absoluteEncryptedFilepath
   */
  public calcRelativeEncryptedFilepath(absoluteEncryptedFilepath: string): string {
    const filepath = relativeOfWorkspace(this.encryptedRootDir, absoluteEncryptedFilepath)
    return filepath.replace(/[/\\]+/g, '/')
  }
}
