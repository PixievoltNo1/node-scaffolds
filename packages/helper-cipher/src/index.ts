export * from './cipher/aes/AesGcmCipher'
export * from './cipher/aes/AesGcmCipherFactory'
export * from './cipher/BaseCipher'
export * from './types/ICipher'
export * from './types/ICipherFactory'
export * from './util'
