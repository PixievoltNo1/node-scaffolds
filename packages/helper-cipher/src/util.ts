import { createHash } from 'node:crypto'

/**
 * Calc Message Authentication Code.
 *
 * @param pieces
 */
export function calcMac(...pieces: Array<Readonly<Buffer>>): Buffer {
  const sha256 = createHash('sha256')
  for (const piece of pieces) sha256.update(piece)
  const mac: Buffer = sha256.digest()
  return mac
}

/**
 * Calc Message Authentication Code from literal string.
 *
 * @param text
 * @param textEncoding
 * @returns
 */
export function calcMacFromString(text: string, textEncoding: BufferEncoding): Buffer {
  const sha256 = createHash('sha256')
  sha256.update(text, textEncoding)
  const mac: Buffer = sha256.digest()
  return mac
}

/**
 * Fill buffer with a random number
 *
 * @param buffer
 * @returns
 */
export function destroyBuffer(buffer: Buffer | null): void {
  if (buffer == null) return
  buffer.fill(0)
  buffer.fill(1)
  buffer.fill(Math.random() * 127)
}

/**
 * Destroy buffers
 *
 * @param buffers
 * @returns
 */
export function destroyBuffers(buffers: Array<Buffer | null> | null): void {
  if (buffers == null) return
  for (const buffer of buffers) destroyBuffer(buffer)
}
