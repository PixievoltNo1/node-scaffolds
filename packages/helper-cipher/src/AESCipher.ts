import type ChalkLogger from '@guanghechen/chalk-logger'
import invariant from '@guanghechen/invariant'
import type { Cipher } from 'crypto'
import crypto from 'crypto'
import { BaseCipher } from './BaseCipher'
import type { ICipher } from './types/cipher'
import { destroyBuffer } from './util/buffer'
import { createRandomIv, createRandomKey } from './util/key'

export interface IAESCipherOptions {
  /**
   *
   */
  logger?: ChalkLogger
  /**
   * @default 'aes-256-gcm'
   */
  algorithm?: crypto.CipherGCMTypes
  /**
   *
   */
  iv?: Buffer
  /**
   *
   */
  key?: Buffer
}

export class AESCipher extends BaseCipher implements ICipher {
  protected readonly ivSize = 32
  protected readonly keySize = 32
  protected readonly algorithm: crypto.CipherGCMTypes
  protected iv: Buffer | null
  protected key: Buffer | null

  constructor(options: IAESCipherOptions = {}) {
    super(options.logger)
    this.algorithm = options.algorithm ?? 'aes-256-gcm'
    this.key = options.key || null
    this.iv = options.iv || null
  }

  public override createSecret(): Buffer {
    const { ivSize, keySize } = this
    const iv: Buffer = createRandomIv(ivSize)
    const key: Buffer = createRandomKey(keySize)
    const secret: Buffer = Buffer.alloc(iv.length + key.length)

    {
      let s = 0
      iv.copy(secret, s, 0, iv.length)
      destroyBuffer(iv)

      s += iv.length
      key.copy(secret, s, 0, key.length)
      destroyBuffer(key)
    }

    return secret
  }

  public override initFromSecret(secret: Readonly<Buffer>): void | never {
    const { ivSize, keySize } = this
    const iv: Buffer = Buffer.alloc(ivSize)
    const key: Buffer = Buffer.alloc(keySize)
    secret.copy(iv, 0, 0, ivSize)
    secret.copy(key, 0, ivSize, ivSize + keySize)

    this.cleanup()
    this.iv = iv
    this.key = key
  }

  public override initFromPassword(password: Readonly<Buffer>): void | never {
    const { ivSize, keySize } = this
    const iv: Buffer = Buffer.alloc(ivSize)
    const key: Buffer = Buffer.alloc(keySize)
    let j = 0

    // generate iv
    for (let i = 0; i < ivSize; ++i) {
      iv[i] = password[j]
      j = (j + 1) % password.length
    }

    // generate key
    for (let i = 0; i < keySize; ++i) {
      key[i] = password[j]
      j = (j + 1) % password.length
    }

    this.cleanup()
    this.iv = iv
    this.key = key
  }

  protected override encipher(): Cipher {
    const { algorithm, key, iv } = this
    invariant(key != null && iv != null, 'NULL_POINTER_ERROR: iv / key is null')

    const encipher = crypto.createCipheriv(algorithm, key, iv)
    return encipher
  }

  public override decipher(): Cipher {
    const { algorithm, key, iv } = this
    invariant(key != null && iv != null, 'NULL_POINTER_ERROR: iv / key is null')

    const decipher = crypto.createCipheriv(algorithm, key, iv)
    return decipher
  }

  public override cleanup(): void {
    destroyBuffer(this.iv)
    destroyBuffer(this.key)
    this.iv = null
    this.key = null
  }
}
