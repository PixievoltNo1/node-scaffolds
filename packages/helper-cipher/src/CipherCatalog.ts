import type { IFilePartItem } from '@guanghechen/helper-file'
import { BigFileHelper, calcFilePartItemsBySize } from '@guanghechen/helper-file'
import { collectAllFilesSync, mkdirsIfNotExists } from '@guanghechen/helper-fs'
import { destroyBuffer } from '@guanghechen/helper-stream'
import invariant from '@guanghechen/invariant'
import crypto from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'
import { CipherPathResolver } from './CipherPathResolver'
import type { ICatalogIndex, ICatalogItem } from './types/catalog'
import type { IFileCipher } from './types/IFileCipher'
import { calcFingerprint, calcMacFromFile } from './util/mac'

export interface ICipherCatalogOptions {
  /**
   * A collect of util funcs for encryption / decryption.
   */
  readonly fileCipher: IFileCipher

  /**
   * Root directory of source files.
   */
  readonly sourceRootDir: string

  /**
   * Root directory of target files.
   */
  readonly targetRootDir: string

  /**
   * Encoding of source contents.
   * @default 'utf8'
   */
  readonly sourceEncoding?: BufferEncoding

  /**
   * Encoding of ciphered catalog index file.
   * @default 'base64'
   */
  readonly cipheredIndexEncoding?: BufferEncoding

  /**
   * Max size (byte) of target file, once the file size exceeds this value,
   * the target file is split into multiple files.
   *
   * For safety, this value should be greater than or equal to 1024.
   *
   * @default Number.MAX_SAFE_INTEGER
   */
  readonly maxTargetFileSize?: number
}

export class CipherCatalog {
  public readonly sourceEncoding: BufferEncoding
  public readonly cipheredIndexEncoding: BufferEncoding
  public readonly maxTargetFileSize: number
  public readonly pathResolver: CipherPathResolver
  protected readonly fileCipher: IFileCipher
  protected readonly fileHelper: BigFileHelper
  protected readonly items: Array<Readonly<ICatalogItem>>
  protected readonly sourceFilepathMap: Map<string, Readonly<ICatalogItem>>
  protected readonly targetFilepathSet: Set<string>
  protected readonly targetPartPathSet: Set<string>
  protected lastCheckTime: string | null

  constructor(options: ICipherCatalogOptions) {
    const sourceRootDir = options.sourceRootDir
    const targetRootDir = options.targetRootDir
    const sourceEncoding = options.sourceEncoding ?? 'utf8'
    const cipheredIndexEncoding = options.cipheredIndexEncoding ?? 'base64'
    const maxTargetFileSize = Math.max(1024, options.maxTargetFileSize ?? Number.MAX_SAFE_INTEGER)

    invariant(
      fs.existsSync(sourceRootDir),
      `[FILEPATH_NOT_FOUND] cannot find sourceRootDir: ${sourceRootDir}`,
    )

    this.pathResolver = new CipherPathResolver({ sourceRootDir, encryptedRootDir: targetRootDir })
    this.fileCipher = options.fileCipher
    this.fileHelper = new BigFileHelper({ encoding: undefined })
    this.sourceEncoding = sourceEncoding
    this.cipheredIndexEncoding = cipheredIndexEncoding
    this.items = []
    this.maxTargetFileSize = maxTargetFileSize
    this.sourceFilepathMap = new Map()
    this.targetFilepathSet = new Set()
    this.targetPartPathSet = new Set()
    this.lastCheckTime = null

    mkdirsIfNotExists(targetRootDir, true)
  }

  /**
   * Load catalog states from ciphered contents.
   * @param cipheredContent
   */
  public load(cipheredContent: string): void {
    const { fileCipher, cipheredIndexEncoding } = this
    const cipherData: Buffer = Buffer.from(cipheredContent, cipheredIndexEncoding)

    // Decrypt data.
    const plainContent = fileCipher.cipher.decrypt(cipherData)

    // Remove salt.
    const content = this.strip(plainContent.toString())

    // parse the content
    const { items, lastCheckTime } = JSON.parse(content) as ICatalogIndex

    const { sourceFilepathMap: sfm, targetFilepathSet: tfs, targetPartPathSet: tps } = this

    this.reset()
    this.lastCheckTime = lastCheckTime
    for (const item of items) {
      this.items.push(item)
      sfm.set(item.sourceFilepath, item)
      tfs.add(item.targetFilename)
      for (const part of item.targetParts) tps.add(part)
    }
  }

  /**
   * Dump catalog data
   */
  public dump(): string {
    const { fileCipher, items, lastCheckTime, sourceEncoding, cipheredIndexEncoding } = this

    const data: ICatalogIndex = { lastCheckTime, items }
    const plaintextContent: string = JSON.stringify(data)

    // Add salt.
    const content = this.seal(plaintextContent)

    // save into the index file
    const sourceData: Buffer = Buffer.from(content, sourceEncoding)
    const cipherData = fileCipher.cipher.encrypt(sourceData).toString(cipheredIndexEncoding)

    destroyBuffer(sourceData)
    return cipherData
  }

  /**
   * Load data from index file.
   * @param indexFilepath   The path of catalog index file
   */
  public async loadFromFile(indexFilepath: string): Promise<void> {
    invariant(
      fs.existsSync(indexFilepath),
      `[FILEPATH_NOT_FOUND] cannot find index file: ${indexFilepath}`,
    )

    // load content from index file
    const { sourceEncoding } = this
    const cipherContent = fs.readFileSync(indexFilepath, { encoding: sourceEncoding })
    this.load(cipherContent)
  }

  /**
   * Dump catalog data and save into the index file.
   * @param indexFilepath   The path of catalog index file
   */
  public async save(indexFilepath: string): Promise<void> {
    const cipherData = this.dump()
    const { sourceEncoding } = this

    mkdirsIfNotExists(indexFilepath, false)
    fs.writeFileSync(indexFilepath, cipherData, { encoding: sourceEncoding })
  }

  /**
   * Reset catalog states.
   */
  public reset(): void {
    this.items.splice(0, this.items.length)
    this.sourceFilepathMap.clear()
    this.targetFilepathSet.clear()
    this.targetPartPathSet.clear()
    this.lastCheckTime = null
  }

  /**
   * Delete invalid entries on the catalog and clean up unregistered target files.
   */
  public cleanup(): void {
    const {
      pathResolver,
      items,
      sourceFilepathMap: sfm,
      targetFilepathSet: tfs,
      targetPartPathSet: tps,
    } = this

    // Delete invalid entries on the catalog.
    for (let i = 0, _size = items.length; i < _size; ) {
      const item = items[i]
      const absoluteSourceFilepath = pathResolver.calcAbsoluteSourceFilepath(item.sourceFilepath)

      if (fs.existsSync(absoluteSourceFilepath)) {
        i += 1
        continue
      }

      _size -= 1
      items.splice(i, 1)
      sfm.delete(item.sourceFilepath)
      tfs.delete(item.targetFilename)
      for (const part of item.targetParts) tps.delete(part)
    }

    // Clean up unregistered target.
    const targetFiles = collectAllFilesSync(pathResolver.encryptedRootDir)
    for (const absoluteTargetFilepath of targetFiles) {
      const part = pathResolver.calcRelativeEncryptedFilepath(absoluteTargetFilepath)
      if (tps.has(part)) continue
      fs.unlinkSync(absoluteTargetFilepath)
    }
  }

  /**
   * Update the lastCheckTime of the workspace.
   * @param date
   */
  public touch(date?: Date): void {
    const mtime = (date ?? new Date()).toISOString()
    if (this.lastCheckTime != null && this.lastCheckTime >= mtime) return
    this.lastCheckTime = mtime
  }

  /**
   * Check if files are damaged.
   */
  public checkIntegrity(): void | never {
    const { pathResolver } = this
    for (const item of this.items) {
      for (const part of item.targetParts) {
        const filepath = pathResolver.calcAbsoluteEncryptedFilepath(part)
        invariant(fs.existsSync(filepath), `[INTEGRITY DAMAGE] cannot found ${filepath}`)
      }
    }

    const allTargetFiles = collectAllFilesSync(pathResolver.encryptedRootDir)
    invariant(
      allTargetFiles.length === this.targetPartPathSet.size,
      `[INTEGRITY DAMAGE] there are ${this.items.length} files in index file,` +
        ` but actually there are ${allTargetFiles.length} files in the cipher directory`,
    )
  }

  /**
   * Register a item into the catalog and perform some cleanup operations.
   *
   * @param _sourceFilepath
   * @returns
   */
  public async register(_sourceFilepath: string): Promise<void> {
    const { pathResolver } = this
    const absoluteSourceFilepath = pathResolver.calcAbsoluteSourceFilepath(_sourceFilepath)
    const sourceFilepath = pathResolver.calcRelativeSourceFilepath(absoluteSourceFilepath)

    const { sourceFilepathMap: sfm, targetFilepathSet: tfs } = this
    const stat: fs.Stats = fs.statSync(absoluteSourceFilepath)
    const size = stat.size
    const mtime = stat.mtime.toISOString()
    const mac = await calcMacFromFile(absoluteSourceFilepath)
    const fingerprint = calcFingerprint(mac)

    let item: ICatalogItem | undefined = sfm.get(sourceFilepath)
    if (item != null) {
      // Return directly if there is no modification.
      if (
        item.size === size &&
        item.fingerprint === fingerprint &&
        item.targetParts.every(p => fs.existsSync(p))
      ) {
        return
      }

      item.mtime = mtime
      item.fingerprint = fingerprint
      item.size = size
    } else {
      const targetFilename = this.generateTargetFilename(sourceFilepath)
      item = {
        fingerprint,
        size,
        mtime,
        sourceFilepath,
        targetFilename,
        targetParts: [],
      }

      this.items.push(item)
      sfm.set(item.sourceFilepath, item)
      tfs.add(item.targetFilename)
    }

    await this.writeTargets(item)
  }

  public async decryptAll(sourceBakRootDir: string): Promise<void> {
    const { pathResolver } = this
    for (const item of this.items) {
      const sourceBakFilepath = path.join(sourceBakRootDir, item.sourceFilepath)
      await this.fileCipher.decryptFiles(
        item.targetParts.map(p => pathResolver.calcAbsoluteEncryptedFilepath(p)),
        sourceBakFilepath,
      )
    }
  }

  /**
   * Check if the file has modified since it was last saved.
   * @param _sourceFilepath
   */
  public isModified(_sourceFilepath: string, _stat?: fs.Stats): boolean {
    if (this.lastCheckTime == null) return true

    const { pathResolver } = this
    const absoluteSourceFilepath = pathResolver.calcAbsoluteSourceFilepath(_sourceFilepath)
    const sourceFilepath = pathResolver.calcRelativeSourceFilepath(absoluteSourceFilepath)

    const item: ICatalogItem | undefined = this.sourceFilepathMap.get(sourceFilepath)
    if (item === undefined) return true

    const stat = _stat ?? fs.statSync(absoluteSourceFilepath)
    return stat.size !== item.size || this.lastCheckTime < stat.mtime.toISOString()
  }

  /**
   * Generate a target filename.
   * @returns
   */
  public generateTargetFilename(sourceFilepath: string): string {
    const { sourceFilepathMap: sfm, targetFilepathSet: tfs } = this
    const item = sfm.get(sourceFilepath)
    if (item != null) return item.targetFilename

    for (;;) {
      const filename: string = crypto.randomBytes(32).toString('hex')
      if (!tfs.has(filename)) return filename
    }
  }

  /**
   * Write one or more target files according to the information recorded by
   * the catalog item.
   *
   * @param item
   */
  protected async writeTargets(item: ICatalogItem): Promise<void> {
    const { pathResolver, fileCipher, fileHelper, targetPartPathSet: tps, maxTargetFileSize } = this

    const absoluteSourceFilepath = pathResolver.calcAbsoluteSourceFilepath(item.sourceFilepath)
    const absoluteTargetFilepath = pathResolver.calcAbsoluteEncryptedFilepath(item.targetFilename)

    // Encrypt source file.
    await fileCipher.encryptFile(absoluteSourceFilepath, absoluteTargetFilepath)

    // Split target file.
    const parts: IFilePartItem[] = calcFilePartItemsBySize(
      absoluteTargetFilepath,
      maxTargetFileSize,
    )
    const partFilepaths: string[] = await fileHelper.split(absoluteTargetFilepath, parts)

    // Remove the original big target file.
    if (partFilepaths.length > 1) fs.unlinkSync(absoluteTargetFilepath)

    // Update target parts.
    const targetParts: string[] = partFilepaths.map(p =>
      pathResolver.calcRelativeEncryptedFilepath(p),
    )
    for (const filepath of item.targetParts) tps.delete(filepath)
    for (const filepath of targetParts) tps.add(filepath)
    for (const filepath of item.targetParts) {
      if (tps.has(filepath) || !fs.existsSync(filepath)) continue
      fs.unlinkSync(filepath)
    }
    // eslint-disable-next-line no-param-reassign
    item.targetParts = targetParts
  }

  /**
   * Adds some random characters for obfuscation.
   * @param content
   */
  protected seal(content: string): string {
    const startSaltSize = Math.ceil(Math.random() * 70 + 30)
    const endSaltSize = Math.ceil(Math.random() * 70 + 30)
    return (
      crypto.randomBytes(startSaltSize).toString('hex') +
      content +
      crypto.randomBytes(endSaltSize).toString('hex')
    )
  }

  /**
   * Remove salts.
   * @param content
   */
  protected strip(content: string): string {
    return content.replace(/^[0-9a-z]+/, '').replace(/[0-9a-z]+$/, '')
  }
}
