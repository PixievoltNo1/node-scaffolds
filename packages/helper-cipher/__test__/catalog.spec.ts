import { collectAllFilesSync } from '@guanghechen/helper-file'
import fs from 'fs-extra'
import { desensitize, locateFixtures, unlinkSync } from 'jest.helper'
import path from 'path'
import { AESCipher, CipherCatalog } from '../src'

function expectEqual(catalog1: CipherCatalog, catalog2: CipherCatalog): void {
  expect(desensitize((catalog1 as any).items)).toEqual(desensitize((catalog2 as any).items))
}

describe('CipherCatalog', function () {
  const password = Buffer.from('@guanghechen/helper-cipher')
  const cipher = new AESCipher()
  cipher.initFromPassword(password)

  describe('preserve', function () {
    test('load / save', async function () {
      const catalog1 = new CipherCatalog({
        cipher,
        sourceRootDir: locateFixtures('catalog/source'),
        targetRootDir: locateFixtures('catalog/target/load-or-save'),
      })

      const catalog2 = new CipherCatalog({
        cipher,
        sourceRootDir: catalog1.sourceRootDir,
        targetRootDir: catalog1.targetRootDir,
      })

      const catalogIndexFilepath = locateFixtures('catalog/catalog.load-or-save.txt')

      catalog1.cleanup()
      const files = collectAllFilesSync(catalog1.sourceRootDir)
      for (const file of files) expect(catalog1.isModified(file)).toEqual(true)
      for (const file of files) await catalog1.register(file)

      catalog1.touch()
      for (const file of files) expect(catalog1.isModified(file)).toEqual(false)

      try {
        expect(fs.existsSync(catalogIndexFilepath)).toBe(false)
        await catalog1.save(catalogIndexFilepath)
        expect(fs.existsSync(catalogIndexFilepath)).toBe(true)

        await catalog2.loadFromFile(catalogIndexFilepath)
        expectEqual(catalog1, catalog2)
      } finally {
        unlinkSync(catalogIndexFilepath)
      }
    })

    test('checkIntegrity', async function () {
      const catalog = new CipherCatalog({
        cipher,
        sourceRootDir: locateFixtures('catalog/source'),
        targetRootDir: locateFixtures('catalog/target/checkIntegrity'),
      })

      catalog.cleanup()

      const tmpSourceFilepath = path.join(catalog.sourceRootDir, 'xx-' + Math.random())
      fs.writeFileSync(tmpSourceFilepath, 'ss')

      expect(collectAllFilesSync(catalog.targetRootDir).length).toEqual(0)
      const sourceFilepaths = collectAllFilesSync(catalog.sourceRootDir)
      for (const file of sourceFilepaths) {
        await catalog.register(file)
      }

      const targetFilepaths = collectAllFilesSync(catalog.targetRootDir)
      expect(targetFilepaths.length).toEqual(3)

      expect(() => catalog.checkIntegrity()).not.toThrow()
      fs.unlinkSync(targetFilepaths[0])
      expect(() => catalog.checkIntegrity()).toThrow('[INTEGRITY DAMAGE] cannot found')

      expect((catalog as any).lastCheckTime).toEqual(null)

      const prevDate = new Date().toISOString()
      catalog.touch()
      const nextDate = new Date().toISOString()
      expect(
        (catalog as any).lastCheckTime != null &&
          prevDate <= (catalog as any).lastCheckTime &&
          nextDate >= (catalog as any).lastCheckTime,
      ).toBe(true)

      for (const file of sourceFilepaths) await catalog.register(file)
      unlinkSync(tmpSourceFilepath)
      expect(collectAllFilesSync(catalog.targetRootDir).length).toEqual(3)
      catalog.cleanup()
      expect(collectAllFilesSync(catalog.targetRootDir).length).toEqual(2)
    })

    test('basic', async function () {
      const catalog1 = new CipherCatalog({
        cipher,
        sourceRootDir: locateFixtures('catalog/source'),
        targetRootDir: locateFixtures('catalog/target/basic'),
      })

      const catalog2 = new CipherCatalog({
        cipher,
        sourceRootDir: catalog1.sourceRootDir,
        targetRootDir: catalog1.targetRootDir,
      })

      const catalogIndexFilepath = locateFixtures('catalog/catalog.basic.txt')
      await catalog1.loadFromFile(catalogIndexFilepath)
      await catalog2.loadFromFile(catalogIndexFilepath)

      expect(catalog1).toEqual(catalog2)
      expect(catalog1.dump()).not.toEqual(catalog2.dump())
      expect(catalog1.dump()).not.toEqual(
        fs.readFileSync(catalogIndexFilepath, catalog1.sourceEncoding),
      )

      const files = collectAllFilesSync(catalog1.sourceRootDir)
      for (const file of files) {
        await catalog2.register(file)
      }

      expectEqual(catalog1, catalog2)
      expect((catalog1 as any).items).toMatchSnapshot()
    })
  })

  describe('big-file', function () {
    test('basic', async function () {
      const catalog = new CipherCatalog({
        cipher,
        sourceRootDir: locateFixtures('catalog/source'),
        targetRootDir: locateFixtures('catalog/target/big-file'),
        maxTargetFileSize: 1024 * 4, // 4KB
      })
      const sourceBakRootDir = locateFixtures('catalog/target/big-file-bak')

      catalog.cleanup()
      const sourceFiles = collectAllFilesSync(catalog.sourceRootDir).sort()
      for (const file of sourceFiles) {
        await catalog.register(file)
      }

      fs.emptyDirSync(sourceBakRootDir)
      await catalog.decryptAll(sourceBakRootDir)

      const bakSourceFiles = collectAllFilesSync(sourceBakRootDir).sort()
      expect(sourceFiles.length).toEqual(bakSourceFiles.length)
      for (let i = 0; i < sourceFiles.length; ++i) {
        expect(fs.readFileSync(sourceFiles[i])).toEqual(fs.readFileSync(bakSourceFiles[i]))
      }
    })
  })
})
