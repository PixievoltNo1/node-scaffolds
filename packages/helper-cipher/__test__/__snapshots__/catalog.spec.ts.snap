// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CipherCatalog preserve basic 1`] = `
Array [
  Object {
    "fingerprint": "51eca7cd4b2e00b0ebea5b1f12c89ba8191d3d22",
    "mtime": "2021-07-03T03:18:08.101Z",
    "size": 10294,
    "sourceFilepath": "1.json",
    "targetFilename": "c8ce553149b9632066e5267c6f1f5efb645b87895a86aceda9a5db86129dcf4c",
    "targetParts": Array [
      "c8ce553149b9632066e5267c6f1f5efb645b87895a86aceda9a5db86129dcf4c",
    ],
  },
  Object {
    "fingerprint": "55a714c90180c22ad3034a1114164a73020138df",
    "mtime": "2021-07-03T03:18:08.101Z",
    "size": 11925,
    "sourceFilepath": "2.json",
    "targetFilename": "230f16fcf9aee9cc5a858e020a9a8d146a70f862ca53dd9e44d048d635a12f6a",
    "targetParts": Array [
      "230f16fcf9aee9cc5a858e020a9a8d146a70f862ca53dd9e44d048d635a12f6a",
    ],
  },
]
`;
